File: ios/TronCoin.swift
//
//  TronCoin.swift
//  coinswallet
//
//  Created by Tal Grynbaum on 30/06/2023.
//

import Foundation
import WalletCore

class TronCoin: CoinFactory.Coin {
    private var addressIndex: Int = 0
    
    override init(mnemonic: String) {
        super.init(mnemonic: mnemonic)
    }
    
    override func getNewAddress(isTestNet:Bool) -> String {
        let privateKey = wallet.getKeyForCoin(coin: .tron)
        let address = CoinType.tron.deriveAddress(privateKey: privateKey)
        addressIndex += 1
        return address
    }
    
    override func getPrivateKey(isTestNet:Bool) -> String {
        let privateKeyBytes = wallet.getKeyForCoin(coin: .tron).data
        return privateKeyBytes.map { String(format: "%02x", $0) }.joined()
    }
  
  override func getDeriveAddresses() -> NSMutableArray {
    let result: NSMutableArray = []
    for i in 0..<50 {
          let derivePath = String(format: "m/44'/195'/0'/%@/0", String(i))
          let privateKey = wallet.getKey(coin: .tron, derivationPath: derivePath)
          let address = CoinType.tron.deriveAddress(privateKey: privateKey)
          let yourAuxDic: NSMutableDictionary = [:]
          yourAuxDic["derivePath"] = derivePath
          yourAuxDic["privateKey"] = privateKey.data.hexString
          yourAuxDic["address"] = address
          result.add(yourAuxDic)
      }
      return result;
    }
  
  override func addCustomDerivation(derivePath:String) -> NSMutableDictionary {
      let privateKey = wallet.getKey(coin: .tron, derivationPath: derivePath)
      let address = CoinType.tron.deriveAddress(privateKey: privateKey)
      let yourAuxDic: NSMutableDictionary = [:]
      yourAuxDic["derivePath"] = derivePath
      yourAuxDic["privateKey"] = privateKey.data.hexString
      yourAuxDic["address"] = address
      return yourAuxDic;
  }

    
    override func signTransaction(rawData: String) -> String {
        // Implement the logic to sign a Tron transaction using the provided raw data
        return ""
    }
}

File: android/app/src/main/java/com/coinswallet/coins/TronCoin.java
// TronCoin.java
package com.coinswallet.coins;

import com.coinswallet.CoinFactory;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;

import wallet.core.jni.HDWallet;
import wallet.core.jni.CoinType;
import wallet.core.jni.PrivateKey;
import java.util.Base64;
import java.nio.charset.StandardCharsets;
import android.util.Log;

public class TronCoin extends CoinFactory.Coin {
    private final String address;

    public TronCoin(String mnemonic) {
        super(mnemonic);
        this.address = CoinType.TRON.deriveAddress(wallet.getKeyForCoin(CoinType.TRON));
    }

    @Override
    public String getNewAddress(Boolean isTestNet) {
        return this.address;
    }

    @Override
    public String getPrivateKey(Boolean isTestNet) {
        byte[] privateKeyBytes = wallet.getKeyForCoin(CoinType.TRON).data();
        String hexPrivateKey = CoinFactory.Coin.toHex(privateKeyBytes);
        System.out.println("hexPrivateKey: " + hexPrivateKey);
        return hexPrivateKey;
        // return new String(privateKeyBytes, StandardCharsets.UTF_8);
    }

    public ReadableArray getDeriveAddresses() {
        WritableArray results = Arguments.createArray();
        for(int i=0; i < 50; i++){
            WritableMap obj = Arguments.createMap();
            String derivePath = "m/44'/195'/0'/" + i + "/0";
            PrivateKey tempPrivateKey = wallet.getKey(CoinType.TRON,derivePath);
            String address = CoinType.TRON.deriveAddress(tempPrivateKey);
            String privateKeyString =  CoinFactory.Coin.toHex(tempPrivateKey.data());
            obj.putString("address",address);
            obj.putString("derivePath",derivePath);
            obj.putString("privateKey",privateKeyString);
            results.pushMap(obj);
        }
        return results;
    }

    @Override
    public ReadableMap addCustomDerivation(String derivePath) {
        WritableMap obj = Arguments.createMap();
        PrivateKey tempPrivateKey = wallet.getKey(CoinType.TRON,derivePath);
        String address = CoinType.TRON.deriveAddress(tempPrivateKey);
        String privateKeyString =  CoinFactory.Coin.toHex(tempPrivateKey.data());
        obj.putString("address",address);
        obj.putString("derivePath",derivePath);
        obj.putString("privateKey",privateKeyString);
        return obj;
    }

    @Override
    public String signTransaction(String rawData) {
        // You need to implement this method according to the specific requirements of
        // Tron transactions.
        return null;
    }
}
